File: ebitengine/states/gameover.go
package states

import (
	"fmt"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/inpututil"
	"github.com/hajimehoshi/ebiten/v2/text/v2"
	"github.com/hajimehoshi/ebiten/v2/vector"
	"github.com/thesimpledev/radianthorizon/core"
	"github.com/thesimpledev/radianthorizon/ui"
)

// GameOver is the win/loss screen state.
type GameOver struct {
	sm    *core.StateManager
	fonts *Fonts

	victory      bool
	score        int
	hp           int
	playAgainBtn *ui.Button
	menuBtn      *ui.Button
}

// NewGameOver creates the game over state.
func NewGameOver(sm *core.StateManager, fonts *Fonts) *GameOver {
	return &GameOver{sm: sm, fonts: fonts}
}

func (go_ *GameOver) Enter(data any) {
	if d, ok := data.(map[string]any); ok {
		if v, ok := d["victory"].(bool); ok {
			go_.victory = v
		}
		if s, ok := d["score"].(int); ok {
			go_.score = s
		}
		if h, ok := d["hp"].(int); ok {
			go_.hp = h
		}
	}

	btnW, btnH := 200.0, 50.0
	btnX := (core.WindowW - btnW) / 2
	btnY := core.WindowH * 0.65

	go_.playAgainBtn = ui.NewButton("Play Again", btnX, btnY, btnW, btnH, go_.fonts.Default, func() {
		go_.sm.Switch("game_play", nil)
	})

	menuY := btnY + btnH + 16
	go_.menuBtn = ui.NewButton("Main Menu", btnX, menuY, btnW, btnH, go_.fonts.Default, func() {
		go_.sm.Switch("menu", nil)
	})
}

func (go_ *GameOver) Exit() {}

func (go_ *GameOver) Update() error {
	mx, my := ebiten.CursorPosition()
	go_.playAgainBtn.HandleMouseMove(mx, my)
	go_.menuBtn.HandleMouseMove(mx, my)

	if inpututil.IsMouseButtonJustPressed(ebiten.MouseButtonLeft) {
		go_.playAgainBtn.HandleMousePress(mx, my)
		go_.menuBtn.HandleMousePress(mx, my)
	}
	if inpututil.IsMouseButtonJustReleased(ebiten.MouseButtonLeft) {
		go_.playAgainBtn.HandleMouseRelease(mx, my)
		go_.menuBtn.HandleMouseRelease(mx, my)
	}

	if inpututil.IsKeyJustPressed(ebiten.KeyEnter) || inpututil.IsKeyJustPressed(ebiten.KeySpace) {
		go_.sm.Switch("game_play", nil)
	}

	return nil
}

func (go_ *GameOver) Draw(screen *ebiten.Image) {
	// Background
	vector.DrawFilledRect(screen, 0, 0, core.WindowW, core.WindowH, core.ColorBG, false)

	// Title
	var title string
	var titleColor = core.ColorRed
	if go_.victory {
		titleColor = core.ColorGold
		title = "VICTORY!"
	} else {
		title = "DEFEATED"
	}
	titleOp := &text.DrawOptions{}
	tw, _ := text.Measure(title, go_.fonts.Title, 0)
	titleOp.GeoM.Translate((core.WindowW-tw)/2, core.WindowH*0.25)
	titleOp.ColorScale.ScaleWithColor(titleColor)
	text.Draw(screen, title, go_.fonts.Title, titleOp)

	// Score
	scoreText := fmt.Sprintf("Score: %d", go_.score)
	scoreOp := &text.DrawOptions{}
	sw, _ := text.Measure(scoreText, go_.fonts.Score, 0)
	scoreOp.GeoM.Translate((core.WindowW-sw)/2, core.WindowH*0.38)
	scoreOp.ColorScale.ScaleWithColor(core.ColorWhite)
	text.Draw(screen, scoreText, go_.fonts.Score, scoreOp)

	// HP
	hpText := fmt.Sprintf("Final HP: %d", go_.hp)
	hpOp := &text.DrawOptions{}
	hw, _ := text.Measure(hpText, go_.fonts.Score, 0)
	hpOp.GeoM.Translate((core.WindowW-hw)/2, core.WindowH*0.38+36)
	hpOp.ColorScale.ScaleWithColor(core.ColorWhite)
	text.Draw(screen, hpText, go_.fonts.Score, hpOp)

	// Buttons
	go_.playAgainBtn.Draw(screen)
	go_.menuBtn.Draw(screen)
}

File: love/states/game_over.lua
-- states/game_over.lua
-- Win/loss screen showing the final score with a "Play Again" button.

local Button       = require("ui.button")
local C            = require("core.constants")
local StateManager = require("core.state_manager")

local GameOver = {}

local title_font = nil
local score_font = nil
local default_font = nil

function GameOver:enter(data)
    if not title_font then
        title_font = love.graphics.newFont(48)
        score_font = love.graphics.newFont(24)
        default_font = love.graphics.newFont(16)
    end

    self.victory = data and data.victory or false
    self.score = data and data.score or 0
    self.hp = data and data.hp or 0

    -- Play Again button
    local btn_w, btn_h = 200, 50
    local btn_x = (C.WINDOW_W - btn_w) / 2
    local btn_y = C.WINDOW_H * 0.65

    self.play_again_btn = Button:new("Play Again", btn_x, btn_y, btn_w, btn_h, function()
        StateManager.switch("game_play")
    end)

    -- Main Menu button
    local menu_y = btn_y + btn_h + 16
    self.menu_btn = Button:new("Main Menu", btn_x, menu_y, btn_w, btn_h, function()
        StateManager.switch("menu")
    end)
end

function GameOver:update(dt)
end

function GameOver:draw()
    -- Background
    love.graphics.setColor(C.COLOR_BG)
    love.graphics.rectangle("fill", 0, 0, C.WINDOW_W, C.WINDOW_H)

    -- Title
    love.graphics.setFont(title_font)
    local title
    if self.victory then
        love.graphics.setColor(C.COLOR_GOLD)
        title = "VICTORY!"
    else
        love.graphics.setColor(C.COLOR_RED)
        title = "DEFEATED"
    end
    local title_w = title_font:getWidth(title)
    love.graphics.print(title, (C.WINDOW_W - title_w) / 2, C.WINDOW_H * 0.25)

    -- Score
    love.graphics.setFont(score_font)
    love.graphics.setColor(C.COLOR_WHITE)
    local score_text = "Score: " .. self.score
    local score_w = score_font:getWidth(score_text)
    love.graphics.print(score_text, (C.WINDOW_W - score_w) / 2, C.WINDOW_H * 0.38)

    -- HP
    local hp_text = "Final HP: " .. self.hp
    local hp_w = score_font:getWidth(hp_text)
    love.graphics.print(hp_text, (C.WINDOW_W - hp_w) / 2, C.WINDOW_H * 0.38 + 36)

    -- Buttons (use default font)
    love.graphics.setFont(default_font)
    self.play_again_btn:draw()
    self.menu_btn:draw()
end

function GameOver:mousepressed(x, y, button)
    self.play_again_btn:mousepressed(x, y, button)
    self.menu_btn:mousepressed(x, y, button)
end

function GameOver:mousereleased(x, y, button)
    self.play_again_btn:mousereleased(x, y, button)
    self.menu_btn:mousereleased(x, y, button)
end

function GameOver:mousemoved(x, y, dx, dy)
    self.play_again_btn:mousemoved(x, y)
    self.menu_btn:mousemoved(x, y)
end

function GameOver:keypressed(key)
    if key == "return" or key == "space" then
        StateManager.switch("game_play")
    end
end

return GameOver
